### PR TITLE
Implement `TestScenario::remove_object_by_id`

### PR DESCRIPTION
**Pull request by lxfind**
When there are multiple objects of same type in the same account address, we need a way retrieve individual objects by Object ID.
This PR adds that.
Ideally we should also add a `can_remove_object_by_id` function, but that seems to me will require make a almost full copy of the function. I could do that, but want to also ask for suggestions/ideas before I do.

This is one of the TODOs in #1324.

### CODE REVIEW

File: sui_programmability/framework/sources/TestScenario.move [hunk 1]
```diff
@@ -142,9 +149,33 @@ module Sui::TestScenario {
...
+            last_tx_start_index(scenario)
+        );
+        let object_opt = Option::none();
+        while (!Vector::is_empty(&objects)) {
+            let element = Vector::pop_back(&mut objects);
+            if (ID::id(&element) == &id) {
+                assert!(Option::is_none(&object_opt), EDUPLICATE_OBJCET_ID_FOUND);
+                Option::fill(&mut object_opt, element);
+            } else {
+                delete_object_for_testing(element);
```
Comment (666lcz) [inline sui_programmability/framework/sources/TestScenario.move:166]: Previously from the naming I assumed this function will remove the object from the "global inventory" on the rust side, but it looks like this is just a [no-op](https://github.com/MystenLabs/sui/blob/317b91464ca631af4713f95ee8a1b79afb9e983f/sui_programmability/framework/src/natives/test_scenario.rs#L302). Maybe we can consider adding some comments  noting that it won't be removed from the inventory or give it another name like `drop_object_for_testing`

File: sui_programmability/framework/sources/TestScenario.move [hunk 2]
```diff
@@ -142,9 +149,33 @@ module Sui::TestScenario {
     /// Same as `remove_object`, but returns the object of type `T` with object ID `id`.
     /// Should only be used in cases where current tx sender has more than one object of
     /// type `T` in their inventory.
-    public fun remove_object_by_id<T: key>(_scenario: &mut Scenario, _id: ID): T {
-        // TODO: implement me
-        abort(100)
+    /// TODO: Perhaps also define can_remove_object_by_id? How to avoid code duplication?
```
Comment (666lcz) [inline sui_programmability/framework/sources/TestScenario.move:152]: I think you can at least extract the first part into a helper function `get_object_by_id`

```
public fun remove_object_by_id<T: key>(scenario: &mut Scenario, id: ID): T {
        let object_opt = get_object_by_id<T>(scenario, id);

        assert!(Option::is_some(&object_opt), EOBJECT_ID_NOT_FOUND);
        let object = Option::extract(&mut object_opt);
        Option::destroy_none(object_opt);

        assert!(!Vector::contains(&scenario.removed, &id), EALREADY_REMOVED_OBJECT);
        Vector::push_back(&mut scenario.removed, id);

        object
    }

    /// Return `true` if a call to `remove_object_by_id<T>(scenario, id)` will succeed
    public fun can_remove_object_by_id<T: key>(scenario: &mut Scenario, id: ID): bool {
        if (Vector::contains(&scenario.removed, &id)) {
            return false
        };
        let object_opt = get_object_by_id(scenario, id);
        if (Option::is_some(&object_opt)) {
            let object = Option::extract<T>(&mut object_opt);
            Option::destroy_none(object_opt);
            delete_object_for_testing(object);
            return true
        };
        Option::destroy_none(object_opt);
        false
    }
    
   fun get_object_by_id<T: key>(scenario: &mut Scenario, id: ID): Option<T> {
        let sender = sender(scenario);
        let objects: vector<T> = get_inventory<T>(
            sender,
            last_tx_start_index(scenario)
        );
        let object_opt = Option::none();
        while (!Vector::is_empty(&objects)) {
            let element = Vector::pop_back(&mut objects);
            if (ID::id(&element) == &id) {
                assert!(Option::is_none(&object_opt), EDUPLICATE_OBJCET_ID_FOUND);
                Option::fill(&mut object_opt, element);
            } else {
                delete_object_for_testing(element);
            }
        };
        Vector::destroy_empty(objects);
        object_opt
    }

```

**Review by 666lcz — APPROVED** [review]
Nice!

**Comment by lxfind** [comment]
Thank you @666lcz !

File: sui_programmability/framework/sources/TestScenario.move (hunk 2 above)
Comment (lxfind) [reply sui_programmability/framework/sources/TestScenario.move:152]: One caveat: `get_object_by_id` should not trigger any assertions though. I will think about it more.

**Pull request merged**